### PR TITLE
Elgg 4 work

### LINKED ISSUE #12225: FR: elgg/lightbox should have a helper function for opening ajax content in a lightbox
some special magic is done when clicking on a link with the class `elgg-lightbox` to open the lightbox and then load the ajax data and reinit the lightbox with the response data.

This works great.

If however want that same magic when i call `lightbox.open` i have to recreate it myself :(

A helper function would be nice

talking about https://github.com/Elgg/Elgg/blob/3.x/views/default/elgg/lightbox.js#L115-L132

### LINKED ISSUE #12558: Revalidate implementation of Collection
Scrutinizer issue
https://scrutinizer-ci.com/g/Elgg/Elgg/issues/3.0/files/engine/classes/Elgg/Collections/Collection.php?selectedSeverities%5B0%5D=10&orderField=path&order=asc&fileId=engine/classes/Elgg/Cli/UpgradeCommand.php&honorSelectedPaths=0

https://github.com/Elgg/Elgg/blob/0f920398c776301badb404f4cb3984d3542f4caf/engine/classes/Elgg/Collections/Collection.php#L272-L282

### LINKED ISSUE #12684: Drop support for container permissions without subtype
Container permission overrides are becoming a nightmare to manage, because some plugins call canWriteToContainer() without type/subtype. In projects with granular permissions checks this turns into a nightmare: following the hooks and their outputs and devising some really hacky ways of managing it all. It would be easier to just not support the generic check.

### PR DESCRIPTION
**Pull request by jeabakker**
feat(webservices): option to enable / disable API keys

chore(access): type/subtype is required for container permissions 
fixes #12684

fix(collections): corrected implementation of SeekableIterator
fixes #12558

feat(js): allow lightbox to load JS dependencies
fixes #12225

### CODE REVIEW

File: mod/web_services/classes/ElggApiKey.php [hunk 1]
```diff
@@ -119,4 +119,36 @@ public function regenerateKeys() {
 		
 		return true;
 	}
+	
+	/**
+	 * Check is the API key is active
+	 *
+	 * @return bool
+	 */
+	public function isKeyActive() {
```
Comment (jdalsem) [inline mod/web_services/classes/ElggApiKey.php:128]: change function name to `hasActiveKeys` to keep plurar usage of the keys word intact

File: mod/web_services/classes/ElggApiKey.php [hunk 2]
```diff
@@ -119,4 +119,36 @@ public function regenerateKeys() {
...
+		
+		return (bool) $keys->active;
+	}
+	
+	/**
+	 * Enables the API key for use by API requests
+	 *
+	 * @return bool
+	 */
+	public function enableKey() {
```
Comment (jdalsem) [inline mod/web_services/classes/ElggApiKey.php:142]: change function name to `enableKeys` to keep plurar usage of the keys word intact

File: mod/web_services/classes/ElggApiKey.php [hunk 3]
```diff
@@ -119,4 +119,36 @@ public function regenerateKeys() {
...
+		return _elgg_services()->apiUsersTable->enableAPIUser($this->getPublicKey());
+	}
+	
+	/**
+	 * Disables the API key for use by API requests
+	 *
+	 * @return bool
+	 */
+	public function disableKey() {
```
Comment (jdalsem) [inline mod/web_services/classes/ElggApiKey.php:151]: change function name to `disableKeys` to keep plurar usage of the keys word intact

File: mod/web_services/languages/en.php [hunk 4]
```diff
@@ -73,6 +75,11 @@
 	'webservices:action:api_key:edit:success' => "API token saved successfully",
 	'webservices:action:api_key:regenerate:success' => "The API keys have been regenerated",
 
+	'webservices:action:api_key:toggle_active:enable:success' => "The API key was successfully enabled",
+	'webservices:action:api_key:toggle_active:enable:error' => "An error occurred while enableing the API key",
```
Comment (jdalsem) [inline mod/web_services/languages/en.php:79]: typo: enabling

File: mod/web_services/languages/en.php [hunk 5]
```diff
@@ -73,6 +75,11 @@
...
+	'webservices:action:api_key:toggle_active:disable:success' => "The API key was successfullt disabled",
```
Comment (jdalsem) [inline mod/web_services/languages/en.php:80]: typo: successfully

File: mod/web_services/languages/en.php [hunk 6]
```diff
@@ -73,6 +75,11 @@
...
+	'webservices:action:api_key:toggle_active:disable:error' => "An error occurred while disableing the API key",
```
Comment (jdalsem) [inline mod/web_services/languages/en.php:81]: typo: disabling

File: mod/web_services/views/default/object/api_key.php [hunk 7]
```diff
@@ -43,11 +43,27 @@
 	$content .= elgg_format_element('div', ['class' => ['hidden', 'webservices-secret-' . $entity->guid]], $secret);
 }
 
+// imprint
+$imprint = [];
+
+if ($entity->isKeyActive()) {
+	$imprint[] = [
+		'icon_name' => 'check',
```
Comment (jdalsem) [inline mod/web_services/views/default/object/api_key.php:51]: rewrite to use a ternary inline if/else function... saves a lot of lines of code.. can do for both icon_name and content

File: mod/web_services/views/default/object/api_key.php [hunk 8]
```diff
@@ -43,11 +43,27 @@
...
+		'content' => elgg_echo('status:disabled'),
+	];
+}
+
 $params = [
 	'title' => $entity->getDisplayName(),
 	'byline' => false,
 	'access' => false,
 	'content' => $content,
+	'imprint' => $imprint,
```
Comment (jdalsem) [inline mod/web_services/views/default/object/api_key.php:66]: if imprint was rewritten to ternary notation, you can add it directly here

**Review by jdalsem — CHANGES_REQUESTED** [review]

File: mod/web_services/views/default/object/api_key.php (hunk 7 above)
Comment (jeabakker) [reply mod/web_services/views/default/object/api_key.php:51]: don't agree

File: mod/web_services/views/default/object/api_key.php (hunk 8 above)
Comment (jeabakker) [reply mod/web_services/views/default/object/api_key.php:66]: don't agree

File: engine/classes/Elgg/Notifications/SubscriptionsService.php [hunk 9]
```diff
@@ -48,25 +56,27 @@ public function __construct(Database $db, array $methods = []) {
 	 * );
 	 *
 	 * @param NotificationEvent $event Notification event
+	 *
 	 * @return array
 	 */
-	public function getSubscriptions(NotificationEvent $event) {
+	public function getSubscriptions(NotificationEvent $event, array $methods = []) {
```
Comment (jdalsem) [inline engine/classes/Elgg/Notifications/SubscriptionsService.php:62]: why default to an empty array?

File: engine/classes/Elgg/Notifications/SubscriptionsService.php [hunk 10]
```diff
@@ -90,18 +101,20 @@ public function getSubscriptions(NotificationEvent $event) {
 	 * );
 	 *
 	 * @param int $container_guid GUID of the entity acting as a container
+	 *
 	 * @return array User GUIDs (keys) and their subscription types (values).
 	 */
-	public function getSubscriptionsForContainer($container_guid) {
-
-		$subscriptions = [];
+	public function getSubscriptionsForContainer(int $container_guid, array $methods = []) {
```
Comment (jdalsem) [inline engine/classes/Elgg/Notifications/SubscriptionsService.php:107]: why default to empty array?

File: engine/classes/Elgg/Notifications/SubscriptionsService.php [hunk 11]
```diff
@@ -150,37 +164,35 @@ public function removeSubscription($userGuid, $method, $targetGuid) {
 	 * @param int $container_guid The GUID of the subscription target
 	 * @return array
 	 */
-	protected function getSubscriptionRecords($container_guid) {
-
-		$container_guid = (int) $container_guid;
-
+	protected function getSubscriptionRecords(int $container_guid, array $methods = []) {
```
Comment (jdalsem) [inline engine/classes/Elgg/Notifications/SubscriptionsService.php:167]: default empty array needed?

File: engine/classes/Elgg/UserCapabilities.php [hunk 12]
```diff
@@ -244,13 +244,13 @@ public function canEditAnnotation(ElggEntity $entity, $user_guid = 0, ElggAnnota
...
 	 * @param ElggEntity $entity    Container entity
-	 * @param int        $user_guid The GUID of the user creating the entity (0 for logged in user).
 	 * @param string     $type      The type of entity we're looking to write
 	 * @param string     $subtype   The subtype of the entity we're looking to write
+	 * @param int        $user_guid The GUID of the user creating the entity (0 for logged in user).
 	 *
 	 * @return bool
 	 */
-	public function canWriteToContainer(ElggEntity $entity, $user_guid = 0, $type = 'all', $subtype = 'all') {
+	public function canWriteToContainer(ElggEntity $entity, string $type, string $subtype, int $user_guid = 0) {
```
Comment (jdalsem) [inline engine/classes/Elgg/UserCapabilities.php:253]: do we need to document the order change? we documented other changes from this class in the upgrade notes

Comment (jeabakker) [reply engine/classes/Elgg/UserCapabilities.php:253]: removed other docs

**Pull request merged**